### PR TITLE
[config] Select optimizers by config type instead of by name

### PR DESCRIPTION
**Pull request by YeonwooSung**
## Summary

Reworked per review: instead of validating `optimizer_name`, optimizer selection is now typed, so an unknown optimizer is unrepresentable rather than rejected at runtime.

- **`Optimizer(Configurable, torch.optim.Optimizer)`** in a new `torchtitan/components/optimizer/base.py`. Its `Config` splits fields into two channels: `to_param_group_kwargs()` for hyperparameters recorded on each PyTorch param group (these may differ per group), and `to_factory_kwargs()` for instance-wide constructor arguments. A subclass names the latter in `_FACTORY_FIELDS`; everything else is per-group.
- **`from_torch(torch_cls)`** adapts `torch.optim` classes; `Adam` and `AdamW` are built with it. It copies the torch class `__name__`, so `_log_optimizer`'s `type(optimizer).__name__` output is unchanged.
- **`DistMuon`** becomes a native `Optimizer` with `compute_sharding_by_fqn` and `bucket_configs` as its factory fields. `build_dist_muon` is gone: its param-group normalization moved into `__init__` and its layout documentation onto the class.
- **`ParamGroupConfig.optimizer: Optimizer.Config`** replaces `optimizer_name` + `optimizer_kwargs`. `OptimizersContainer` now batches param groups by optimizer *class* rather than by name string, preserving one instance per optimizer type, and builds each through its config.

Removed: `optimizer_name`, `optimizer_kwargs`, `optimizer_factory_kwargs_by_name`, and the name-to-factory table. Dropping that table also removes the `flex_shard` import from `optimizer.py`, which is what lets `dist_muon` import the `Optimizer` base without an import cycle.

### Notes on two constraints this had to satisfy

**The frozen CLI.** `ParamGroupConfig.optimizer` is annotated `tyro.conf.Suppress`. `_config_types` in `test_no_new_cli_options.py` expands subclasses of a declared component base, so an unsuppressed field would add one option per field per optimizer subclass (`optimizer.param_groups.optimizer.lr`, `.momentum`, `.ns_coefficients`, ...) and fail `test_no_new_options`. This is the same escape hatch `Trainer.Config.model_spec` uses. The three removed fields come out of `_FROZEN_CLI_OPTIONS`; nothing is added, and the freeze assertion only rejects additions.

**Behaviour preservation.** `to_param_group_kwargs()` emits every field, whereas the old dicts let unset keys fall through to constructor defaults, so the `Config` defaults are set to match the previous constructor defaults exactly (covered by a test). `DistMuon.Config.foreach` defaults to `False` because `_validate_groups` rejects a truthy value and Kimi K2.7 sets `implementation="foreach"` for the whole container; the container still merges as `{**impl_kwargs, **to_param_group_kwargs()}`, so a config field overrides the global setting exactly as `optimizer_kwargs` did.

Most call sites were unaffected: roughly a hundred registry entries go through `default_adamw()`, so only its body changed. `qwen3` and `kimi_k2_7` construct `ParamGroupConfig` directly and were updated, along with the Kimi routed-expert layout alignment hook, which now finds the `DistMuon.Config` in `param_groups` instead of a name-keyed dict (building a fresh list, since `dataclasses.replace` is a shallow copy).

## Test plan

- [x] New `tests/unit_tests/cpu/test_optimizer_base.py`: the two-channel split, `from_torch` naming, `AdamW`/`DistMuon` config defaults matching the previous constructor defaults, and `DistMuon.Config.foreach` overriding a global `foreach`.
- [x] `test_optimizer_param_groups.py`: the `optimizer_name` validation tests are replaced by batching tests (two `AdamW` groups with different `lr` land in one instance with two param groups; `Adam` + `AdamW` get separate instances) and a factory-kwarg disagreement test.
- [ ] CI CPU unit tests, and the GPU suites for `test_qk_clip` and `flex_shard/test_dist_muon`.

I could not run the suites locally (no GPU, and the local torch is too old to import torchtitan), so the class mechanics were verified separately: MRO linearisation of `from_torch`, `_owner` wiring, `super(Optimizer, self).__init__` reaching `torch.optim.Optimizer`, `param_names` surviving onto the group, and that `tyro.conf.Suppress` is both necessary and sufficient for the freeze test. CI is the real check.

Happy to split this differently, or hand it back if you would rather own the design.

### CODE REVIEW

File: torchtitan/components/optimizer/optimizer.py [hunk 1]
```diff
@@ -60,12 +68,20 @@ class ParamGroupConfig:
     E.g. '.*bias$', '.*norm.*', '.*\\.embed_tokens\\..*', '.*' (catch-all)"""
 
     optimizer_name: str
```
Comment (tianyu-l) [inline torchtitan/components/optimizer/optimizer.py]: instead, we should
- create torchtitan native `Optimizer(Configurable, torch.optim.Optimizer)`
- create torchtitan native `AdamW` and `DistMuon` (https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/flex_shard/dist_muon.py#L252) -- the `AdamW` one can be done via a `from_torch` helper.
- here specify `optimizer: Optimizer.Config`

if it sounds complicated, leave it to me.

Comment (YeonwooSung) [reply torchtitan/components/optimizer/optimizer.py]: Took the design rather than leaving it to you -- pushed in 3e4121c. If you would rather own it, say so and I will close this; no attachment to my version.

Shape follows your three bullets. `Optimizer(Configurable, torch.optim.Optimizer)` lives in a new `components/optimizer/base.py`, `from_torch` builds `Adam`/`AdamW`, `DistMuon` is native, and `ParamGroupConfig` takes `optimizer: Optimizer.Config`.

The one thing I had to invent: `Config` needs two channels, because the container batches param groups by optimizer and one instance can carry several groups with different `lr`. So `Optimizer.Config` exposes `to_param_group_kwargs()` (recorded per group) and `to_factory_kwargs()` (instance-wide, passed once). A subclass names the latter in `_FACTORY_FIELDS`; `DistMuon` uses it for `compute_sharding_by_fqn` and `bucket_configs`, which is what let `optimizer_factory_kwargs_by_name` go away. Batching now keys on the config class, and groups sharing a class must agree on the factory kwargs.

Two things worth your eye:

1. `ParamGroupConfig.optimizer` is `tyro.conf.Suppress`. Without it `_config_types` expands subclasses of the declared base and adds `optimizer.param_groups.optimizer.{lr,betas,momentum,ns_coefficients,...}` to the CLI, failing `test_no_new_options`. Suppressing matches `Trainer.Config.model_spec`, but it does mean the optimizer can no longer be set from the command line -- only from a config registry. Tell me if that trade is wrong and I will find another way.

2. Removing `build_dist_muon` is load-bearing, not cleanup. `optimizer.py` imported it for the name-to-factory table; `dist_muon` now imports the `Optimizer` base, so keeping that table would close an import cycle.

Behaviour: `Config` defaults are pinned to the previous constructor defaults, since `to_param_group_kwargs()` emits every field where the old dicts let unset keys fall through. `DistMuon.Config.foreach=False` preserves the Kimi override against `implementation="foreach"`.

I could not run the suites (no GPU here, and the local torch is too old to import torchtitan), so I verified the mechanics standalone -- MRO, `_owner` wiring, `super(Optimizer, self).__init__` reaching torch, `param_names` surviving onto the group, and that Suppress is necessary and sufficient. CI is the actual check, so please treat the test plan checkbox as unproven until it goes green.
